tests/Application/Actions/Hooks/StripeChargeUpdateTest.php

```diff
@@ -367,6 +428,73 @@ public function testSuccessfulFullRefund(): void
         $time = (string) time();
 
+        $chatterProphecy = $this->prophesize(StripeChatterInterface::class);
+
+        // Double-check that the normal success case isn't messaging Slack.
+        $chatterProphecy->send(Argument::type(ChatMessage::class))->shouldNotBeCalled();
+
+        $donationRepoProphecy = $this->prophesize(DonationRepository::class);
+
+        $donationRepoProphecy
+            ->findOneBy(['chargeId' => 'ch_externalId_123'])
+            ->willReturn($donation)
+            ->shouldBeCalledOnce();
+
+        $donationRepoProphecy
+            ->releaseMatchFunds($donation)
+            ->shouldBeCalledOnce();
+
+        $donationRepoProphecy
+            ->push(Argument::type(Donation::class), false)
+            ->willReturn(true)
+            ->shouldBeCalledOnce();
+
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+        $container->set(StripeChatterInterface::class, $chatterProphecy->reveal());
+
+        // act
+        $request = $this->createRequest('POST', '/hooks/stripe', $body)
+            ->withHeader('Stripe-Signature', $this->generateSignature($time, $body, $webhookSecret));
+
+        $response = $app->handle($request);
+
+        // assert
+        $this->assertEquals('ch_externalId_123', $donation->getChargeId());
+        $this->assertEquals(DonationStatus::Refunded, $donation->getDonationStatus());
+        $this->assertEquals('1.00', $donation->getTipAmount());
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    /**
+     * Verifies that:
+     * * Slack gets a warning with the correct details
+     * * The donation record is updated with the new status, and
+     * * Match funds are released.
+     */
+    public function testOverRefundSendsSlackNoticeAndUpdatesRecord(): void
+    {
+        // arrange
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        $body = $this->getStripeHookMock('ch_over_refunded');
+        $donation = $this->getTestDonation();
+        $webhookSecret = $container->get('settings')['stripe']['accountWebhookSecret'];
+        $time = (string) time();
+
+        $chatterProphecy = $this->prophesize(StripeChatterInterface::class);
+
+        $options = (new SlackOptions())
+            ->block((new SlackHeaderBlock('[test] Over-refund detected')))
+            ->block((new SlackSectionBlock())->text('Over-refund detected for donation 12345678-1234-1234-1234-1234567890ab based on charge.refunded hook. Donation inc. tip was 124.45 GBP and refund or dispute was 134.45 GBP'))
+            ->iconEmoji(':o');
+        $message = (new ChatMessage('Over-refund detected'))
+            ->options($options);
```

Comment:
Could rename `$message` to `$expectedMessage` to make it clearer - but I think even better is just to inline all this stuff so it appears only in the assert section, e.g.:

```php
        $chatterProphecy->send((new ChatMessage('Over-refund detected'))
            ->options((new SlackOptions())
                ->block((new SlackHeaderBlock('[test] Over-refund detected')))
                ->block((new SlackSectionBlock())->text('Over-refund detected for donation 12345678-1234-1234-1234-1234567890ab based on charge.refunded hook. Donation inc. tip was 124.45 GBP and refund or dispute was 134.45 GBP'))
                ->iconEmoji(':o')))->shouldBeCalledOnce();
```

Also TIL that prophecy will let you do `shouldBeCalled` after the the action, using the revelation as a spy instead of a mock. Nice!

Reply:
Yeah, in theory that would be nicer. I tried moving some of the calls to `// assert` but of course if they need to define things about the prophesised method's behaviour in the same chained call then to do it with the fluent interface you need both things before you `// act`

Reply:
I think it was moving donation repo calls at the same time which broke things, but I didn't dig into it further